controllers/storageclassclaim/storageclassclaim_controller.go

```diff
@@ -92,20 +139,317 @@ func (r *StorageClassClaimReconciler) Reconcile(ctx context.Context, request rec
+		pvList := corev1.PersistentVolumeList{}
+		err := r.list(&pvList, client.MatchingFields{
+			"spec.storageClassName": r.storageClassClaim.Name,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to list PersistentVolumes: %s", err)
+		}
+		if len(pvList.Items) > 0 {
+			return fmt.Errorf("StorageClass %s is still in use by one or more PV(s)", r.storageClassClaim.Name)
+		}
+
+		// Remove finalizer from StorageClass.
+		storageClass.SetFinalizers(remove(storageClass.Finalizers, StorageClassFinalizer))
+		if err := r.update(storageClass); err != nil {
+			return fmt.Errorf("failed to remove finalizer from StorageClass: %s", err)
+		}
+
+		// Delete the StorageClass.
+		err = r.delete(existing)
+		if err != nil {
+			r.log.Error(err, "Failed to delete StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+			return err
+		}
+	}
+	r.log.Info("Creating StorageClass.", "StorageClass", klog.KRef(storageClass.Namespace, existing.Name))
+	err := r.Client.Create(r.ctx, storageClass)
+	if err != nil {
+		return fmt.Errorf("failed to create StorageClass: %v", err)
+	}
+	return nil
+}
+
+func (r *StorageClassClaimReconciler) reconcileProviderPhases() (reconcile.Result, error) {
+	log := ctrllog.FromContext(r.ctx)
 	log.Info("Running StorageClassClaim controller in Converged/Provider Mode")
 	return reconcile.Result{}, nil
 }
 
 func (r *StorageClassClaimReconciler) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
-		For(&ocsv1alpha1.StorageClassClaim{}).
+		For(&ocsv1alpha1.StorageClassClaim{}, builder.WithPredicates(predicate.Funcs{
+			UpdateFunc: func(e event.UpdateEvent) bool {
+				// Validation check.
+				oldStorageClassClaim := e.ObjectOld.(*ocsv1alpha1.StorageClassClaim)
+				newStorageClassClaim := e.ObjectNew.(*ocsv1alpha1.StorageClassClaim)
+				newStorageClassClaimKRef := klog.KRef(newStorageClassClaim.Namespace, newStorageClassClaim.Name)
+				switch {
+				case oldStorageClassClaim.Spec.Type != newStorageClassClaim.Spec.Type:
+					r.log.Info("StorageClassClaim type change not allowed, skipping update",
+						"StorageClassClaim", newStorageClassClaimKRef)
+					return false
+				case oldStorageClassClaim.Spec.EncryptionMethod != newStorageClassClaim.Spec.EncryptionMethod:
+					r.log.Info(
+						"StorageClassClaim encryption method change not allowed, skipping update",
+						"StorageClassClaim", newStorageClassClaimKRef)
+					return false
+				default:
+					return true
+				}
+			},
+		})).
+		Owns(&storagev1.StorageClass{}).
 		Complete(r)
 }
+
+func (r *StorageClassClaimReconciler) get(obj client.Object) error {
+	key := client.ObjectKeyFromObject(obj)
+	return r.Client.Get(r.ctx, key, obj)
+}
+
+func (r *StorageClassClaimReconciler) list(obj client.ObjectList, listOptions ...client.ListOption) error {
+	if len(listOptions) == 0 {
+		return r.Client.List(r.ctx, obj, client.InNamespace(r.namespace))
+	}
+	return r.Client.List(r.ctx, obj, listOptions[0])
```

Comment:
Two concerns:
1. Why are we making assumptions about the namespace in a general util? 
This will prevent usage of the util for listing cluster-wide resources.

I would prefer to just specify the namespace option whenever we call the function
`r.list(&list, client.InNamespace(r.namespace))`

2. `Client.List` can accept any number of list options, why are we only passing the first one?

Taking the two comments into account, I would suggest
```suggestion
func (r *StorageClassClaimReconciler) list(obj client.ObjectList, listOptions ...client.ListOption) error {
	return r.Client.List(r.ctx, obj, listOptions...)
```

Reply:
ACK